Add alluxio reliability doc (#6506)
NVIDIA/spark-rapids: docs/get-started/getting-started-alluxio.md

- [Running Spark on Alluxio](https://docs.alluxio.io/os/user/stable/en/compute/Spark.html)
- [Performance Tuning](https://docs.alluxio.io/os/user/stable/en/operation/Performance-Tuning.html )
- [Alluxio troubleshooting](https://docs.alluxio.io/os/user/stable/en/operation/Troubleshooting.html)

## Alluxio reliability
The properties mentioned in this section can be found in [Alluxio configuration](https://docs.alluxio.io/os/user/stable/en/reference/Properties-List.html)

> ### Avoid long time response if masters or workers hang

Reviewer: ```suggestion
### Dealing with Client side delays in response from master or workers 
```

The default value of `alluxio.user.rpc.retry.max.duration` is 2 minutes.  

> If master is done or hang, the `alluxio.user.rpc.retry.max.duration` will cause at least 2 minutes waiting, it's too long.

Reviewer: ```suggestion
If the master is not responding, possibly due to it crashing or GC pause , `alluxio.user.rpc.retry.max.duration` will cause the client to retry for 2 minutes.  This is a very long time and can cause delays in the running job, so we suggest lowering this value to 10 seconds.
```

The default value of `alluxio.user.block.read.retry.max.duration` is 5 minutes.

> If worker is done or hang, the `alluxio.user.block.read.retry.max.duration` will cause 5 minutes waiting, it's too long.

Reviewer: ```suggestion
If the worker is not responding, possibly due to it crashing or GC pause , `alluxio.user.block.read.retry.max.duration` will cause the client to retry for 5 minutes.  This is a very long time and can cause delays in the running job, so we suggest lowering this value to 10 seconds.
```

Reviewer: Actually this one worries me a bit, the definition is "This duration controls for how long Alluxio clients should try reading a single block. ".  So this implies to me it includes the time you are streaming the read.  So if you just have a very large block and are reading from say s3, this time seems low to me.  I would rather see this set a bit higher then 10 seconds. maybe 60 seconds.

Reviewer: as the comment says the chunk size is configurable.  In my opinion there is a trade off here of failing quickly vs once in a while getting something slow downloading.  Failing to fast on something like this in my opinion could be bad, especially if it tries again and fails again and again. There is a reason they have the defaults higher and since this includes the time to actually get a block, I would rather see this higher  60 seconds. We can tune it more later if we see issues but I don't want to arbitrarily cause issues because its to small.

Author: Yes， this explanation is very reasonable, changed to 1 minute.

See also:
```
alluxio.user.rpc.retry.max.duration
alluxio.user.rpc.retry.max.sleep	
alluxio.user.rpc.retry.base.sleep	

alluxio.user.block.read.retry.max.duration
alluxio.user.block.read.retry.sleep.max		
alluxio.user.block.read.retry.sleep.base 		
```
Above configurations define the `ExponentialTimeBoundedRetry` retry policies and `max durations`, we can adjust them to appropriate values.  

Set these properties on Spark because Spark invokes Alluxio client.

> vi $SPARK_HOME/conf/spark-defaults.conf

Reviewer: I would rather say something more generic here about adding hte following confs because user could add through cli with --conf

Author: Done

```
spark.driver.extraJavaOptions    -Dalluxio.user.rpc.retry.max.duration=10sec -Dalluxio.user.block.read.retry.max.duration=10sec
spark.executor.extraJavaOptions  -Dalluxio.user.rpc.retry.max.duration=10sec -Dalluxio.user.block.read.retry.max.duration=10sec
```

> ### Worker Timeout to fail fast

Reviewer: so this header the last header seem very similar, I think perhaps the difference is client vs server configuration, correct?  If so lets add something about that to the title

Reviewer: Perhaps like "Worker server tunings to fail fast"

- [Running Spark on Alluxio](https://docs.alluxio.io/os/user/stable/en/compute/Spark.html)
- [Performance Tuning](https://docs.alluxio.io/os/user/stable/en/operation/Performance-Tuning.html )
- [Alluxio troubleshooting](https://docs.alluxio.io/os/user/stable/en/operation/Troubleshooting.html)

## Alluxio reliability
### Auto start Alluxio master and workers  
It's better to add the following crontab jobs to avoid unexpected closes on Alluxio master or workers.  

> Add a crontab job on the Alluxio maser node to automatically start Alluxio maser:  

Reviewer 2: It seems odd to me to use crontab job to start Alluxio master every minute to fix the master down issue.

Reviewer: I agree, I would rather see us use systemd as well.

Note "master" misspelled above as "maser".

    # try to start Alluxio worker every minute
    crontab -e
    * * * * * ${ALLUXIO_HOME}/bin/alluxio-start.sh -N worker > /tmp/cron-worker.log 2>&1
    ```

The `-N` in above commands means do not try to kill previous running processes before starting new ones. 

### Auto fallback to external file system transparently if Alluxio workers have failures
`alluxio.user.client.cache.timeout.duration` means:  
The timeout duration for local cache I/O operations (reading/writing/deleting). When this property is a positive value,local cache operations after timing out will fail and fallback to external file system but transparent to applications; when this property is a negative value, this feature is disabled.
The default value is -1 which means disable the fallback.  
The following configures set the fallback time out to 10 seconds.
```
vi $SPARK_HOME/conf/spark-defaults.conf

spark.driver.extraJavaOptions    -Dalluxio.user.client.cache.timeout.duration=10sec
spark.executor.extraJavaOptions  -Dalluxio.user.client.cache.timeout.duration=10sec  
```

### Avoid long time response if master or workers have failures
The default value of `alluxio.user.rpc.retry.max.duration` is 2 minutes.  

> If master is done the `alluxio.user.rpc.retry.max.duration` will cause at least 2 minutes waiting, it's also too long.

Author: yes.

- [Running Spark on Alluxio](https://docs.alluxio.io/os/user/stable/en/compute/Spark.html)
- [Performance Tuning](https://docs.alluxio.io/os/user/stable/en/operation/Performance-Tuning.html )
- [Alluxio troubleshooting](https://docs.alluxio.io/os/user/stable/en/operation/Troubleshooting.html)

## Alluxio reliability
The properties mentioned in this section can be found in [Alluxio configuration](https://docs.alluxio.io/os/user/stable/en/reference/Properties-List.html)
### Auto start Alluxio masters and workers

> After installed Alluxio masters and workers, use this [script](../../scripts/alluxio/alluxio-systemd.sh) to create systemd services and start masters and workers.

Reviewer: ```suggestion
After installing Alluxio masters and workers, use this [script](../../scripts/alluxio/alluxio-systemd.sh) to create systemd services and start masters and workers.
```

Author: Done

Reviewer: so I think I would still like to see a small section in this doc about reliability and using something like systemd to manage the processes. This way if users don't use our docker image they can set up themselves.  We don't need the full script and instructions perhaps just a general statement about using systemd services

#### Prerequisites

> - Format Alluxio on masters: 

Reviewer: ```suggestion
- Format Alluxio on the master: 
```

    ```bash
    $ ./bin/alluxio format
    ```
    Or you have executed `alluxio-start.sh master` which internally invoke `alluxio format`  

> - Close all the Alluxio masters and workers

Reviewer: what does this mean?  stop all of the masters and workers?

Author: Will move the systemd script to the docker image. 
Wil test this when creating the docker image

The default value of `alluxio.user.rpc.retry.max.duration` is 2 minutes.  
If master is exited, the `alluxio.user.rpc.retry.max.duration` will cause at least 2 minutes waiting, it's too long.
The default value of `alluxio.user.block.read.retry.max.duration` is 5 minutes.
If worker is exited, the `alluxio.user.block.read.retry.max.duration` will cause 5 minutes waiting, it's too long.

See also:
```
alluxio.user.rpc.retry.max.duration
alluxio.user.rpc.retry.max.sleep	
alluxio.user.rpc.retry.base.sleep	

alluxio.user.block.read.retry.max.duration
alluxio.user.block.read.retry.sleep.max		
alluxio.user.block.read.retry.sleep.base 		
```
Above configurations define the `ExponentialTimeBoundedRetry` retry policies and `max durations`, we can adjust them to appropriate values.  

Set these properties on Spark because Spark invokes Alluxio client.
vi $SPARK_HOME/conf/spark-defaults.conf
```
spark.driver.extraJavaOptions    -Dalluxio.user.rpc.retry.max.duration=10sec -Dalluxio.user.block.read.retry.max.duration=10sec
spark.executor.extraJavaOptions  -Dalluxio.user.rpc.retry.max.duration=10sec -Dalluxio.user.block.read.retry.max.duration=10sec
```

> ### Kick out dead worker early to avoid failures

Reviewer: I think I would just call this Worker Timeout to fail fast.

Do we know the failure scenario's here?  would the worker be dead and thus systemd restarted or have we seen cases worker just hangs?   The reason I ask is there are failing fast and then intermittent issues... like say worker was Gc'ing and then we time it out.  I assume it can join back in?

Author: > Do we know the failure scenario's here?
I tested the killing process.

> would the worker be dead and thus systemd restarted or have we seen cases worker just hangs? 
Until now does not see other scenarios that worker hangs.
Long time GC obviously will cause the worker to hang. 

> The reason I ask is there are failing fast and then intermittent issues... like say worker was Gc'ing and then we time it out. I assume it can join back in?

Yes, If a worker fails fast and then comes back, it can join again.
